ocaml/xapi/repository_helpers.ml

```diff
@@ -1361,15 +1324,139 @@ let prune_updateinfo_for_livepatches livepatches updateinfo =
+  | Guidance.RebootHostOnKernelLivePatchFailure ->
+      true
+  | _ ->
+      false
+
+type pending_ops = {
+    host_get: unit -> Guidance.t list
+  ; host_add:
+         [ `reboot_host
+         | `reboot_host_on_livepatch_failure
+         | `restart_toolstack
+         | `reboot_host_on_xen_livepatch_failure
+         | `reboot_host_on_kernel_livepatch_failure ]
+      -> unit
+  ; host_remove:
+         [ `reboot_host
+         | `reboot_host_on_livepatch_failure
+         | `restart_toolstack
+         | `reboot_host_on_xen_livepatch_failure
+         | `reboot_host_on_kernel_livepatch_failure ]
+      -> unit
+  ; vms_get: unit -> (string * Guidance.t list) list
+  ; vm_add: string -> [`restart_device_model | `restart_vm] -> unit
+  ; vm_remove: string -> [`restart_device_model | `restart_vm] -> unit
+}
+
+let get_ops_of_pending ~__context ~host ~kind =
+  let get_pending_guidances_of_host ~get =
+    get ~__context ~self:host
+    |> List.map (fun g -> Guidance.of_pending_guidance g)
+  in
+  let get_pending_guidances_of_vms ~get =
+    Db.Host.get_resident_VMs ~__context ~self:host
+    |> List.map (fun self -> (self, Db.VM.get_record ~__context ~self))
+    |> List.filter_map (fun (ref, record) ->
+           match
+             ( record.API.vM_is_control_domain
+             , record.API.vM_power_state
+             , Helpers.has_qemu_currently ~__context ~self:ref
+             )
+           with
+           | false, `Running, true | false, `Paused, true ->
+               Some ref
+           | _ ->
+               None
+       )
+    |> List.map (fun vm_ref ->
+           let pending_guidances =
+             get ~__context ~self:vm_ref
+             |> List.map Guidance.of_pending_guidance
+           in
+           (Ref.string_of vm_ref, pending_guidances)
+       )
+  in
+  match kind with
+  | Guidance.Mandatory ->
+      let host_get () =
+        get_pending_guidances_of_host ~get:Db.Host.get_pending_guidances
+      in
+      let host_add value =
+        Db.Host.add_pending_guidances ~__context ~self:host ~value
+      in
+      let host_remove value =
+        Db.Host.remove_pending_guidances ~__context ~self:host ~value
+      in
+      let vms_get () =
+        get_pending_guidances_of_vms ~get:Db.VM.get_pending_guidances
+      in
+      let vm_add vm value =
+        Db.VM.add_pending_guidances ~__context ~self:(Ref.of_string vm) ~value
+      in
+      let vm_remove vm value =
+        Db.VM.remove_pending_guidances ~__context ~self:(Ref.of_string vm)
+          ~value
+      in
+      {host_get; host_add; host_remove; vms_get; vm_add; vm_remove}
+  | _ ->
+      raise Api_errors.(Server_error (internal_error, ["Not implemented kind"]))
+
+let set_pending_guidances ~ops ~coming =
+  let pending_of_host =
+    ops.host_get () |> List.filter (fun x -> not (is_livepatch_failure x))
+  in
+  let to_be_added, to_be_removed =
+    merge_with_pending_guidances ~pending:pending_of_host ~coming
+  in
```

Comment:
As only new XenCenter can use this new version of xapi, then XenCenter should finish applying any pending mandatory guidances if there is any, when XC start to apply update, I think there should have no pending mandatory guidances on any host?
Then when will we need to merge coming mandatory ones with pending? Is this needed only when those pending guidance are from the original `host.pending_guidances` for recommended guidances of old xapi? Then should XC also ensure those legacy guidances got applied before start applying a new update?

Reply:
This is a general logic for all kinds of guidance, including `Recommended` and `Full` which will be added in following PR.